src/storage/recovery/recovery_manager.cpp

```diff
@@ -0,0 +1,934 @@
+                         new_tuple_slot, staged_record->Delta(), true /* insert */);
+    TERRIER_ASSERT(staged_record->GetTupleSlot() == new_tuple_slot,
+                   "Insert should update redo record with new tuple slot");
+    // Create a mapping of the old to new tuple. The new tuple slot should be used for future updates and deletes.
+    tuple_slot_map_[old_tuple_slot] = new_tuple_slot;
+  } else {
+    auto new_tuple_slot = tuple_slot_map_[redo_record->GetTupleSlot()];
+    redo_record->SetTupleSlot(new_tuple_slot);
+    // Stage the write. This way the recovery operation is logged if logging is enabled
+    auto staged_record = txn->StageRecoveryWrite(record);
+    bool result UNUSED_ATTRIBUTE = sql_table_ptr->Update(txn, staged_record);
+    TERRIER_ASSERT(result, "Buffered changes should always succeed during commit");
+  }
+}
+
+void RecoveryManager::ReplayDeleteRecord(transaction::TransactionContext *txn, LogRecord *record) {
+  auto *delete_record = record->GetUnderlyingRecordBodyAs<DeleteRecord>();
+  // Get tuple slot
+  auto new_tuple_slot = GetTupleSlotMapping(delete_record->GetTupleSlot());
+  auto sql_table_ptr = GetSqlTable(txn, delete_record->GetDatabaseOid(), delete_record->GetTableOid());
+
+  // Stage the delete. This way the recovery operation is logged if logging is enabled
+  txn->StageDelete(delete_record->GetDatabaseOid(), delete_record->GetTableOid(), new_tuple_slot);
+
+  // Fetch all the values so we can construct index keys after deleting from the sql table
+  auto initializer = sql_table_ptr->InitializerForProjectedRow(sql_table_ptr->GetAllOids());
+  auto *buffer = common::AllocationUtil::AllocateAligned(initializer.ProjectedRowSize());
+  auto pr = initializer.InitializeRow(buffer);
+  sql_table_ptr->Select(txn, new_tuple_slot, pr);
+
+  // Delete from the table
+  bool result UNUSED_ATTRIBUTE = sql_table_ptr->Delete(txn, new_tuple_slot);
+  TERRIER_ASSERT(result, "Buffered changes should always succeed during commit");
+
+  // Delete from the indexes
+  UpdateIndexesOnTable(txn, delete_record->GetDatabaseOid(), delete_record->GetTableOid(), sql_table_ptr,
+                       new_tuple_slot, pr, false /* delete */);
+  // We can delete the TupleSlot from the map
+  tuple_slot_map_.erase(delete_record->GetTupleSlot());
+  delete[] buffer;
+}
+
+void RecoveryManager::UpdateIndexesOnTable(transaction::TransactionContext *txn, catalog::db_oid_t db_oid,
+                                           catalog::table_oid_t table_oid,
+                                           common::ManagedPointer<storage::SqlTable> table_ptr,
+                                           const TupleSlot &tuple_slot, ProjectedRow *table_pr, const bool insert) {
+  auto db_catalog_ptr = GetDatabaseCatalog(txn, db_oid);
+
+  // Stores ptr to index
+  std::vector<common::ManagedPointer<storage::index::Index>> indexes;
+
+  // Stores index schemas, used to get indexcol_ids
+  std::vector<catalog::IndexSchema> index_schemas;
```

Comment:
This should be a `std::vector<const catalog::IndexSchema &>` to avoid deep copies of the internal expressions.

Reply:
You can't have a container of references. I also can't store pointers to the schema because taking a pointer of a reference (the catalog only returns a reference) forces the pointer to be `const`. The pointer can't be const because `GetIndexedColOids` is not a const function due to the caching we do. I can change IndexSchema to do the logic of `GetIndexedColOids` at construction, that way `GetIndexedColOids` can be const.

Reply:
> I can change IndexSchema to do the logic of `GetIndexedColOids` at construction, that way `GetIndexedColOids` can be const.

Decided to do this.